Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.cpp

```diff
@@ -0,0 +1,243 @@
+        m_input_valid = false;
+    }
+
+    auto maybe_project_path = get_project_full_path();
+
+    if (maybe_project_path.has_value()) {
+        m_full_path_label->set_text(maybe_project_path.value());
+    } else {
+        m_full_path_label->set_text("Invalid name or creation directory.");
+        m_input_valid = false;
+    }
+
+    m_ok_button->set_enabled(m_input_valid);
+}
+
+Optional<String> NewProjectDialog::get_available_project_name()
+{
+    auto create_in = m_create_in_input->text();
+    auto chosen_name = m_name_input->text();
+
+    // Ensure project name isn't empty or entirely whitespace
+    if (chosen_name.is_empty() || chosen_name.is_whitespace())
+        return {};
+
+    // Validate project name with validity regex
+    if (!s_project_name_validity_regex.has_match(chosen_name))
+        return {};
+
+    if (!Core::File::exists(create_in) || !Core::File::is_directory(create_in))
+        return {};
+
+    // Check for up-to 999 variations of the project name, in case it's already taken
+    for (int i = 0; i < 1000; i++) {
+        auto candidate = (i == 0)
+            ? chosen_name
+            : String::formatted("{}-{}", chosen_name, i);
+
+        if (!Core::File::exists(String::formatted("{}/{}", create_in, candidate)))
+            return candidate;
+    }
+
+    return {};
+}
+
+Optional<String> NewProjectDialog::get_project_full_path()
+{
+    // Do not permit forward-slashes in project names
+    if (m_name_input->text().contains("/"))
+        return {};
+
+    auto create_in = m_create_in_input->text();
+    auto maybe_project_name = get_available_project_name();
+
+    if (!maybe_project_name.has_value()) {
+        return {};
+    }
+
+    auto project_name = maybe_project_name.value();
+    auto full_path = LexicalPath(String::formatted("{}/{}", create_in, project_name));
+
+    // Do not permit otherwise invalid paths.
+    if (!full_path.is_valid())
+        return {};
+
+    return full_path.string();
+}
+
+void NewProjectDialog::do_create_project()
+{
+    auto project_template = selected_template();
+    if (!project_template) {
+        GUI::MessageBox::show_error(this, "Could not create project: no template selected.");
+        return;
+    }
+
+    auto maybe_project_name = get_available_project_name();
+    auto maybe_project_full_path = get_project_full_path();
+    if (!maybe_project_name.has_value() || !maybe_project_full_path.has_value()) {
+        GUI::MessageBox::show_error(this, "Could not create project: invalid project name or path.");
+        return;
+    }
+
+    if (project_template->create_project(maybe_project_name.value(), maybe_project_full_path.value())) {
+        // Succesfully created, attempt to open the new project
+        m_created_project_path = maybe_project_full_path.value();
+        done(ExecResult::ExecOK);
+    } else {
+        GUI::MessageBox::show_error(this, "Could not create project: internal error. Please check your parameters and try again.");
+    }
```

Comment:
How about making `project_template` return an `AK::Result<String, String>` instead, with the first being the created project full path (for easier unwrapping below), and the second being a more detailed error string ? "Internal Error" is not as descriptive as we could be, and it seems like `create_project` could return a better message pretty easily.

Reply:
I may very well be missing something obvious, but it seems it's impossible to specify an error value for a `Result` with the same value and error type, which stops us from passing through the unwrapped project full path. It is possible to construct a `Result<void, String>` however, still allowing us to pass through an error string. It doesn't seem to be much functionally differently from a `Optional<string>`, but it seems like a more explicit way to pass an error than just using an `Optional`.